src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -74,12 +76,22 @@ public async Task SendRequestBodyAsync()
                 {
                     using (Http2WriteStream writeStream = new Http2WriteStream(this))
                     {
-                        await _request.Content.CopyToAsync(writeStream).ConfigureAwait(false);
+                        try
+                        {
+                            await _request.Content.CopyToAsync(writeStream, null, cancellationToken).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            // We must handle cancellation before disposing the stream, so that we can make a decision about
```

Comment:
This seems redundant with the catch in the caller

Reply:
The issue here is the `using` statement. Http2WriteStream.Dispose currently sends an empty data frame with the END_STREAM header set, unless the request was cancelled. So if a cancellation occurs during CopyToAsync we'll send the END_STREAM header and then send a RST_STREAM. This code makes sure we cancel first, and don't send the extra frame.

Reply:
Ah, I understand. Hmm, let me think about this.

Reply:
That sounds reasonable to me. I'll give it a try, and see if things end up cleaner.

Reply:
This is done in the latest commit.